Script to debug memtest86plus (efi-version) with GDB in QEMU

koenigr opened the pull request:
When I was working on my own memtest-project, it was very usefull to retrace what happens in the code by debug it with gdb in qemu. I think this could help other memtest-developers to comprehend how to load the symbol-tables at the correct loading address.
This solution contains a script (debug_memtest.sh) with the whole procedure and some modifications at the Makefile and memtest_shared.lds to create a debug symbol table. More explanations and suggestions for further improvements can be found in HOW_TO_DEBUG_WITH_GDB.md

debrouxl commented:
Until now, I have always tested memtest86+ on real computers, and I knew it was possible to use QEMU, but I didn't have a procedure for that... so I consider this to be useful documentation and scripting to have :)

martinwhitaker commented:
I've used qemu (and VirtualBox) a lot during the development of PCMemTest and Memtest86+, but I find it quicker and easier to insert a few print calls than to use gdb. I did try using qemu + gdb to track down one particularly stubborn bug in the startup code, but as gdb didn't track the changes between real/protected/long mode, it wasn't that helpful. But I'm not against adding this script if other people will find it useful.

martinwhitaker commented on build64/Makefile:
```diff
@@ -66,6 +213,8 @@ APP_OBJS = app/badram.o \
 OBJS = boot/startup.o boot/efisetup.o $(SYS_OBJS) $(LIB_OBJS) $(TST_OBJS) $(APP_OBJS)
 
+debug: clean memtest.debug memtest.efi
+
```
The `debug` target should come after the `all` target, to keep `all` as the default.

koenigr replied:
You are right. I have changed it

martinwhitaker commented on build64/Makefile:
```diff
@@ -12,6 +14,151 @@ endif
+# In Progress. This option must not be activated as the right flags are not chosen yet.
+ifeq ($(DEBUG_OPT), 1)
+  OPT = -Og
+  OPT_SYSTEM = \
```
I agree with @debrouxl , we need to make the code compile with -Og instead of doing this.

martinwhitaker commented on build64/ldscripts/memtest_shared.lds:
```diff
@@ -52,5 +52,4 @@ SECTIONS {
 		. = ALIGN(256);
 		_end = .;
 	}
-	/DISCARD/ : { *(*) }
```
This adds ~12KB to the final binary size even without the -Og option, so I'd rather it wasn't done on a non-debug build.

debrouxl replied:
It's due to .eh_frame and .comment, the latter being tiny. To keep a single linker script, they could still quit being discarded, but be stripped out individually by `objcopy`.

martinwhitaker commented on build64/Makefile:
```diff
@@ -66,6 +213,8 @@ APP_OBJS = app/badram.o \
 
 OBJS = boot/startup.o boot/efisetup.o $(SYS_OBJS) $(LIB_OBJS) $(TST_OBJS) $(APP_OBJS)
 
+debug: clean memtest.debug memtest.efi
```
Do you really want to do a clean build every time?

koenigr replied:
Thank you for the remark. It was indeed not the intention to do a clean build every time. I just added it at the end to make sure that the whole script runs from the beginning without errors. I have removed it now

debrouxl commented:
Two more ideas:
* many code bases build with debugging flags, then strip / split the debugging information after linking. There's a build time penalty in doing so, but the slowdown shouldn't be _that_ bad for a freestanding C code base like memtest86+, as opposed to some C++ code bases in which every single file includes dozens of thousands of lines of complicated non-modular headers. I once significantly optimized the build of a code base by switching from convenient `#include <QtCore>` to `#include <QString> #include <QList>...` - just the select few classes that the code actually used. That code base had a lower build-to-test time than the mere fact of installing a new memtest86+ build and rebooting the physical server to test memtest86+.
* an alternate way to write $(OPT) would be to use
```
ifeq ($(DEBUG_S), 1)
  OPT_SMALL=-Og
  OPT_FAST=-Og
else
  OPT_SMALL=-Os
  OPT_FAST=-O3
endif
```
and use these instead of hard-coded -Os / -O3 conditionally overridden by $(OPT) later.

koenigr commented:
Thank you for all of your feedback. I will react to the single comments /suggestions step by step during the next days

koenigr commented:
I have adapted your suggestions in the following way:

1. Build with level -Og
I added a preprocessor directive -DDEBUG_GDB to CFLAGS. If it is defined (in a debug build), then memset and memcpy are implemented in string.c. Otherwise they are used as builtin functions. So there are no changes in a non-debug build.

2. Supported terminals
Now, x-terminal-emulator, gnome-terminal and xterm are supported. If the user wishes to use an other terminal (to not to be forced to install one the supported three), it is possible to pass an own terminal by ./debug_memtest.sh -t <terminal> <exec_cmd>.
There is also a help information callable by ./debug_memtest.sh -t help

3. Discard sections
It seems that it does not help to strip .eh_frame and .comment manually. I still have a big difference in the size. I just have some rough ideas but can't really explain why it is as it is. So the new solution is that I let /DISCARD/ in the original linkerscript and create a temporal linkerscript without /DISCARD/ when executing a debug build. So here is also no change to the original non-debug build

Open tasks:
1. One question is if it would be better to add the files created by the script to "make clean" instead of deleting them by "./debug_memtest.sh -c"

2. As soon as there are no more suggestions for improvement, I would finish the documentation/comments and resolve some small remaining "todos" in the code

debrouxl commented on HOW_TO_DEBUG_WITH_GDB.md:
```diff
@@ -0,0 +1,99 @@
+
+###	Alternatives to gnome-terminal
+
+Todo: add alternatives to gnome-terminal, if gnome-terminal is not installed
```
This is partially fixed now.

koenigr replied:
I have made some little changes to this file. As for me, it is the final version now

debrouxl commented on build64/debug_memtest.sh:
```diff
@@ -0,0 +1,201 @@
+
+Check() {
+	# Check if QEMU and OVMF is installed
+	if ! dpkg -l qemu-system* > /dev/null 2>&1; then
```
Wouldn't a check based on `command -v` be enough here ? That would be much more portable.

koenigr replied:
Yes, that's right. I have added this some time before the other checks. Now all checks are consistent

debrouxl commented on build64/debug_memtest.sh:
```diff
@@ -0,0 +1,201 @@
+		exit 1
+	fi
+
+	if ! dpkg -l ovmf > /dev/null 2>&1; then
```
Checking for the existence of *.fd files in /usr/share/OVMF would avoid querying a package manager and therefore gain portability. AFAICS, all of the Debian, CentOS / RHEL / Fedora and Alpine packages contain files matching that pattern.

koenigr replied:
this is also fixed now

koenigr commented:
I have marked this pull request as ready for review. Let me know if there are any more suggestions for improvement

debrouxl commented:
Works for me after manually rebasing the combined diff, which yields the following diff:
```
--- 177.diff.orig       2023-01-17 12:27:19.433828139 +0100
+++ 177.diff    2023-01-17 12:31:15.919836680 +0100
@@ -8,7 +8,7 @@
  
 -# Generated file
 +# Generated files
- githash.h
+ build_version.h
 +gdbscript
 +memtest_shared_debug.lds
  
@@ -171,7 +171,7 @@
  -include boot/efisetup.d
  -include $(subst .o,.d,$(SYS_OBJS))
  -include $(subst .o,.d,$(LIB_OBJS))
-@@ -87,27 +107,27 @@ boot/%.s: ../boot/%.S ../boot/boot.h
+@@ -86,27 +106,27 @@ boot/%.s: ../boot/%.S ../boot/boot.h
  
  boot/efisetup.o: ../boot/efisetup.c
        @mkdir -p boot
@@ -198,14 +198,14 @@
 -      $(CC) -c $(CFLAGS) -O3 $(INC_DIRS) -o $@ $< -MMD -MP -MT $@ -MF $(@:.o=.d)
 +      $(CC) -c $(CFLAGS) -O3 $(INC_DIRS)  -o $@ $< -MMD -MP -MT $@ -MF $(@:.o=.d)
  
- app/%.o: ../app/%.c app/githash.h
+ app/%.o: ../app/%.c app/build_version.h
        @mkdir -p app
 -      $(CC) -c $(CFLAGS) -Os $(INC_DIRS) -o $@ $< -MMD -MP -MT $@ -MF $(@:.o=.d)
 +      $(CC) -c $(CFLAGS) $(OPT_SMALL)  $(INC_DIRS)  -o $@ $< -MMD -MP -MT $@ -MF $(@:.o=.d)
  
- app/githash.h: FORCE
+ app/build_version.h: FORCE
        @mkdir -p app
-@@ -126,13 +146,18 @@ FORCE:
+@@ -127,13 +147,18 @@ FORCE:
  # Link it statically once so I know I don't have undefined symbols and
  # then link it dynamically so I have full relocation information.
  
```

Looks good enough for me after you add a newline at the end of build64/debug_memtest.sh :)

Passing SMP flags to `qemu-system-x86_64` works as it should. I noticed that the debugger not catch execution when passing `-cpu host -accel kvm` to `qemu-system-x86_64`, but this should not prevent from merging this PR.

Technically, we _could_ move -ggdb3 to CFLAGS to unconditionally build debug info, and strip the debug sections for the non-debug binaries, with a build time penalty... but most of the time, memtest86+ is not being debugged, so it's usually a waste of developer time.

Sam, Martin ?

debrouxl commented on build64/debug_memtest.sh:
```diff
@@ -0,0 +1,222 @@
+# Run QEMU and launch second terminal,
+# wait for connection via gdb
+$TERMINAL gdb -x $GDB_FILE &
+$QEMU $QEMU_FLAGS -s -S
```
Newline at EOF ?

debrouxl commented on build64/debug_memtest.sh:
```diff
@@ -0,0 +1,222 @@
+	GDB_FILE="gdbscript"
+
+	# Check if gdbscript exists. If not, create one.
+    if [ ! -f $GDB_FILE ]
```
This script is indented by a mix of tabs and spaces.

debrouxl commented on lib/string.c:
```diff
@@ -43,6 +43,17 @@ int memcmp(const void *s1, const void *s2, size_t n)
+#ifdef DEBUG_GDB
+void *memcpy (void *dest, const void *src, size_t len)
+{
+  char *d = dest;
```
4 spaces indentation, like in the rest of the code, please :)

debrouxl commented on lib/string.c:
```diff
@@ -125,4 +146,4 @@ char *itoa(int num, char *str)
 
     return str;
-}
+}
```
Newline at EOF, please, though newer GCC versions do no longer warn about the lack thereof :)

koenigr commented:
Hi,

I have rebased this fork onto the last state original memtest86plus:main branch. So there should be no merge-conflicts. Additionally, I have corrected the inconsisted identation in the script.

To this comment:
> Ah, I noticed one usability issue: code breakpoints work on the relocated addresses, but data addresses do not. For instance, I ran `p acpi_config`: that printed the memory at 0x29e40, instead of 0x229e40.
I have made numerous test and I realized the following:

Indeed, there was a little part missing. To the command  `add-symbol-file` I have add the addresses of the respective data sections `-s .data $DATAOFFSET"`

BUT:
It only shows you the address with the first loaded Offset. You can switch both lines
`        echo "add-symbol-file memtest.debug $OFFSET -s .data $DATAOFFSET" >> $GDB_FILE
        echo "add-symbol-file memtest.debug $RELOCADDR -s .data $RELOCDATA" >> $GDB_FILE` 
to either get the variable addresses at the loading address 0x200000 or 0x400000

AND a much bigger "BUT":
the given example with acpi_config is a very special case. It is declared as extern and initialized to {0,0,0.....,false}. I absolutley have no clue for the reason and how to solve it, but it seems to be not possible to get addresses for this combination.

Try it out:
set the line 92 of acpi.c from 
`acpi_t acpi_config = {0, 0, 0, 0, 0, 0, 0, false};`
to
`acpi_t acpi_config = {1, 0, 0, 0, 0, 0, 0, false};`

and voila ... you get an address with the offset

koenigr commented:
Could you please check, if the additions for 32-bit support work on our system/environment or if I have forgotten to include something? When I install ovmf-ia32 on my system, I get OVMF32_CODE_4M.secboot.fd instead of OVMF32_CODE.fd, and what I get seems to not work with the debug_memtest.sh script...

debrouxl commented:
Exactly, users of the Debian ovmf-ia32 packages get an incomplete set of edk2 UEFI BIOS images, where only the secure boot version is available. Last time I had a look at the matter, it seemed that a simple change in debian/rules ought to fix this (adding entries to the "OVMF32_IMAGES  :=" line), but I need to check for real.

debrouxl commented:
Martin, Sam: any feedback before I merge this ? I wouldn't have been able to work on NUMA support or the badram reporting fix using pure physical hardware.

x86fr commented:
I have really no experience working with GDB on Memtest86+, so if you think it's ready to merge without any impact on real-world platforms needing tests, feel free to do it.

debrouxl commented:
Common code is untouched by this PR. The addition and usage of `OPT_SMALL`, `OPT_FAST` and `MS_LDS` in the Makefiles are no-ops, and when `DEBUG_GDB` is not defined, the changes to `lib/string.*` are no-ops as well. 
Also, FWIW, the SHA-256 hashes of the binaries produced before and after this change are identical on my Debian sid amd64 environment - after hex-editing the string containing the Git commit hash prefix to the same value on both sides, that is.

A triple-check is always welcome nevertheless :)

martinwhitaker commented:
Finally found some time to look at this. A couple of comments:

1. I think HOW_TO_DEBUG_WITH_GDB.md should live in the doc directory, not the top level.
2. The script uses OVMF.fd, OVMF_CODE.fd, and OVMF_VARS.fd. On my system only the latter two exist. Searching for an explanation, I found https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=764918#47, which explains that OVMF.fd is just the concatenation of OVMF_CODE.fd and OVMF_VARS.fd, and states "In addition, the unified OVMF.fd should not be provided at all, if possible; it only leads to confusion.". FWIW I just use `-bios /usr/share/edk2/ovmf/OVMF_CODE.fd` when using qemu to debug memtest86+ and ignore OVMF_VARS.fd.
